include/picongpu/plugins/xrayScattering/XrayScattering.kernel

```diff
@@ -0,0 +1,207 @@
+                            cellsGrid,
+                            linearIdx + linAccessBlockBegin
+                        )
+                    );
+
+                    DataSpace< simDim > const
+                        cellGlobalPosition( cellPosition + globalOffset );
+                    floatD_X fieldGlobalPosition = precisionCast< float_X >
+                        ( cellGlobalPosition ) + fieldPos( )[ 0 ];
+                    fieldGlobalPosition *= cellSize.shrink< simDim >( );
+
+                    float_X density = densityBoxGPU( cellPosition )[ 0 ];
+                    // Save the cell position in the beam coordinate system.
+                    float3_X  position_b = probingBeam.coordinateTransform(
+                        currentStep,
+                        fieldGlobalPosition
+                    );
+                    float_X beamFactor = probingBeam( position_b );
+                    // Store the cell value of the function that has to be
+                    // Fourier transformed.
+                    densities[ linearIdx ] = density * beamFactor
+                        / totalSimulationCells;
+                    // Store position in the beam comoving system
+                    // 3rd component is not needed anymore since q_z = 0.
+                    positions[linearIdx] = position_b.shrink< DIM2 >( );
+
+                    // Wait for all threads on the block to finish.
+                    cupla::__syncthreads( acc );
+
+                    // Calculate the density fourier transform:
+                    // Loop over q-vectors in frequency space:
+                    // Each worker process every numWorkers vector.
+                    for(
+                        uint32_t qLoopIdx = workerIdx;
+                        qLoopIdx < totalNumVectors;
+                        qLoopIdx += numWorkers
+                        )
+                    {
+                        float2_X q = scatteringVectors[ qLoopIdx ];
+                        complex_X amplitude( 0.0 );
+                        // Loop over all previously loaded cells:
+                        // This is a volume integral over the local domain
+                        // in the beam coordinate system.
+                        for(
+                            uint32_t rLoopIdx = 0;
+                            rLoopIdx < blockSize;
+                            rLoopIdx++
+                            )
+                        {
+                            float_X dotqr;
+                            density = densities[ rLoopIdx ];
+                            float2_X position = positions[ rLoopIdx ];
+                            dotqr = algorithms::math::dot(
+                                position,
+                                q
+                            );
+                            dotqr *= -1.0;
+                            amplitude += algorithms::math::euler(
+                                density,
```

Comment:
These math functions are now in `pmacc::math`

Reply:
Btw I am not sure how it builds for you with this version. I noticed this due to a build error, also `sincos` in another file needs the same fix.

Reply:
Ha! So it looks like I'm compiling from the wrong source directory. Though I'm not sure why.  This could be the cause of all this! @sbastrakov